apps/shipments/serializers/shipment.py

```diff
@@ -184,7 +188,30 @@ def create(self, validated_data):
             if 'device' in self.context:
                 extra_args['device'] = self.context['device']
 
-            return Shipment.objects.create(**validated_data, **extra_args)
+            if 'tags' in validated_data:
+                shipment_tags = validated_data.pop('tags')
+                shipment = Shipment.objects.create(**validated_data, **extra_args)
+                for shipment_tag in shipment_tags:
+                    ShipmentTag.objects.create(
+                        tag_type=shipment_tag['tag_type'],
+                        tag_value=shipment_tag['tag_value'],
+                        shipment_id=shipment.id,
+                        owner_id=shipment.owner_id
+                    )
+            else:
+                shipment = Shipment.objects.create(**validated_data, **extra_args)
+
+            return shipment
+
+    def validate_tags(self, tags):
+        if len(tags) != len([set(t) for t in {tuple(d.items()) for d in tags}]):
+            raise serializers.ValidationError('Tags field cannot contain duplicates')
+
+        for tag_dict in tags:
```

Comment:
Avoid iterating over the same list twice back-to-back. These validations can be combined in the same loop.